From 9f875ac885717dc8c5e7bd6cd02ca7b65d6f35b4 Mon Sep 17 00:00:00 2001
From: Reto Schneider <code@reto-schneider.ch>
Date: Sun, 2 Dec 2018 06:14:27 +0000
Subject: [PATCH] jtag: sysfsgpio: clean up swd gpios

---
 src/jtag/drivers/sysfsgpio.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/jtag/drivers/sysfsgpio.c b/src/jtag/drivers/sysfsgpio.c
index 5535c71d8..176f252a0 100644
--- a/src/jtag/drivers/sysfsgpio.c
+++ b/src/jtag/drivers/sysfsgpio.c
@@ -561,6 +561,8 @@ static void cleanup_all_fds(void)
 	cleanup_fd(tdo_fd, tdo_gpio);
 	cleanup_fd(trst_fd, trst_gpio);
 	cleanup_fd(srst_fd, srst_gpio);
+	cleanup_fd(swclk_fd, swclk_gpio);
+	cleanup_fd(swdio_fd, swdio_gpio);
 }
 
 static bool sysfsgpio_jtag_mode_possible(void)
-- 
2.11.0

