From 19988e50defdd54da72e8c99ef6f7a07288a13b6 Mon Sep 17 00:00:00 2001
From: Lukas Woodtli <lukas.woodtli@husqvarnagroup.com>
Date: Thu, 2 Nov 2023 14:11:17 +0100
Subject: [PATCH] packaging: Exclude lwm2mserver IPSO definitions from Python
 package

---
 lwm2mserver/setup.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lwm2mserver/setup.py b/lwm2mserver/setup.py
index e100ee3..b6ea86b 100644
--- a/lwm2mserver/setup.py
+++ b/lwm2mserver/setup.py
@@ -37,7 +37,7 @@ setup(
     ],
     python_requires=">=3",
     packages=["lwm2mserver", "lwm2mserver_registry", "wakaama"],
-    package_data={"lwm2mserver": ["py.typed"], "lwm2mserver_registry": ["ipso_registry/definitions/*.xml"]},
+    package_data={"lwm2mserver": ["py.typed"]},
     entry_points={
         "console_scripts": ["lwm2mserver=lwm2mserver:main"],
     },
-- 
2.34.1

