From 2ebe2cf1defba063198e49194826ff7d125c42dc Mon Sep 17 00:00:00 2001
From: Marc Lasch <marc.lasch@husqvarnagroup.com>
Date: Wed, 13 Sep 2023 15:03:24 +0200
Subject: [PATCH] packaging: Exclude lwm2mserver IPSO definitions from Python
 package

---
 lwm2mserver/setup.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lwm2mserver/setup.py b/lwm2mserver/setup.py
index 1e7f7f6..578a90c 100644
--- a/lwm2mserver/setup.py
+++ b/lwm2mserver/setup.py
@@ -37,7 +37,7 @@ setup(
     ],
     python_requires=">=3",
     packages=["lwm2mserver", "lwm2mserver_registry", "wakaama"],
-    package_data={"lwm2mserver": ["py.typed"], "lwm2mserver_registry": ["definitions/*.xml"]},
+    package_data={"lwm2mserver": ["py.typed"]},
     entry_points={
         "console_scripts": ["lwm2mserver=lwm2mserver:main"],
     },
-- 
2.34.1

