From d73ecb58b5ac99a64287ff65a0bad08b7bf73c5a Mon Sep 17 00:00:00 2001
From: Andrej Gessel <andrej.gessel@husqvarnagroup.com>
Date: Tue, 4 Dec 2018 15:33:02 +0000
Subject: [PATCH] Fix missing CMSPAR definition in glibc

Signed-off-by: Andrej Gessel <andrej.gessel@husqvarnagroup.com>
---
 linux-serial-test.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/linux-serial-test.c b/linux-serial-test.c
index f6aca5e..4373a7c 100644
--- a/linux-serial-test.c
+++ b/linux-serial-test.c
@@ -15,6 +15,11 @@
 #include <linux/serial.h>
 #include <errno.h>
 
+/* mips doesn't defined CMSPAR in termios.h */
+#if !defined CMSPAR && defined __mips__
+#define CMSPAR	010000000000
+#endif
+
 // command line args
 int _cl_baud = 0;
 char *_cl_port = NULL;
-- 
2.11.0

