From ca9cedcdec6a5d67d316e3652345dca1b0d28196 Mon Sep 17 00:00:00 2001
From: Michael Zimmermann <michael.zimmermann@grandcentrix.net>
Date: Thu, 21 Jun 2018 13:18:43 +0200
Subject: [PATCH] Clients: fix misleading indentation warning

---
 Clients/dns-sd.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/Clients/dns-sd.c b/Clients/dns-sd.c
index 8cd55c1..70a6989 100644
--- a/Clients/dns-sd.c
+++ b/Clients/dns-sd.c
@@ -1224,9 +1224,9 @@ static void DNSSD_API addrinfo_reply(DNSServiceRef sdref, const DNSServiceFlags
         const unsigned char       *b  = (const unsigned char *      )&s6->sin6_addr;
         if (!if_indextoname(s6->sin6_scope_id, if_name))
             snprintf(if_name, sizeof(if_name), "<%d>", s6->sin6_scope_id);
-            snprintf(addr, sizeof(addr), "%02X%02X:%02X%02X:%02X%02X:%02X%02X:%02X%02X:%02X%02X:%02X%02X:%02X%02X%%%s",
-            b[0x0], b[0x1], b[0x2], b[0x3], b[0x4], b[0x5], b[0x6], b[0x7],
-            b[0x8], b[0x9], b[0xA], b[0xB], b[0xC], b[0xD], b[0xE], b[0xF], if_name);
+        snprintf(addr, sizeof(addr), "%02X%02X:%02X%02X:%02X%02X:%02X%02X:%02X%02X:%02X%02X:%02X%02X:%02X%02X%%%s",
+        b[0x0], b[0x1], b[0x2], b[0x3], b[0x4], b[0x5], b[0x6], b[0x7],
+        b[0x8], b[0x9], b[0xA], b[0xB], b[0xC], b[0xD], b[0xE], b[0xF], if_name);
     }
 
     //go through this only if you have a dnssec validation status
-- 
2.19.2

