From 4d6421c01c7efda6147048e4d5350708de2dd99c Mon Sep 17 00:00:00 2001
From: Michael Zimmermann <michael.zimmermann@grandcentrix.net>
Date: Fri, 8 Jan 2021 15:02:44 +0100
Subject: [PATCH] mDNSPosix: NULL-check ifa_addr

---
 mDNSPosix/mDNSPosix.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mDNSPosix/mDNSPosix.c b/mDNSPosix/mDNSPosix.c
index 43b2014..cdd9663 100755
--- a/mDNSPosix/mDNSPosix.c
+++ b/mDNSPosix/mDNSPosix.c
@@ -1354,7 +1354,7 @@ mDNSlocal int SetupInterfaceList(mDNS *const m)
         struct ifaddrs *i;
         for (i = intfList; i; i = i->ifa_next)
         {
-            if (     ((i->ifa_addr->sa_family == AF_INET)
+            if (     i->ifa_addr && ((i->ifa_addr->sa_family == AF_INET)
 #if HAVE_IPV6
                       || (i->ifa_addr->sa_family == AF_INET6)
 #endif
-- 
2.30.0

