From 509f67aea9218b089cbfc2db9db98f200438ec6c Mon Sep 17 00:00:00 2001
From: Marc Lasch <marc.lasch@husqvarnagroup.com>
Date: Sun, 29 Nov 2020 20:54:38 +0100
Subject: [PATCH] Link libatomic

---
 build_pynng.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/build_pynng.py b/build_pynng.py
index cc89e20..400e950 100644
--- a/build_pynng.py
+++ b/build_pynng.py
@@ -24,7 +24,7 @@ if sys.platform == 'win32':
 else:
     objects = ['./nng/build/libnng.a', "./mbedtls/prefix/lib/libmbedtls.a",
                "./mbedtls/prefix/lib/libmbedx509.a", "./mbedtls/prefix/lib/libmbedcrypto.a"]
-    libraries = ['pthread']
+    libraries = ['pthread', 'atomic']
 
 
 ffibuilder.set_source(
-- 
2.25.1

